### PR TITLE
Rewrite Perspective Jupyterlab and Phosphor API

### PR DESCRIPTION
**Pull request by sc1f**
This PR completely rewrites the Jupyterlab and Phosphor API for Perspective. 

Using the new python API, perspective-jupyterlab now uses `Table` and `PerspectiveManager` to handle data loads, view creation, etc. The viewer in the widget has access to a proxy table, and all operations in the viewer are proxied to the underlying Table instance in the kernel. Thus, the new widget is bound to the semantics of Table, and accepts any of the data formats accepted by `perspective.Table`.

As part of this PR, the jupyterlab and phosphor widgets have been rewritten. Specifically, the Phosphor widget is made inert, i.e. `load` must be called on the widget in order to provide the viewer with data. The Jupyterlab widget extends the phosphor widget and implements a client that interacts with `PerspectiveManager` in the kernel. 

Example notebooks have been added to illustrate semantics and usage of the new widgets in JupyterLab.

### CODE REVIEW

File: packages/perspective-jupyterlab/src/ts/view.ts [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+        this.model.on('change:sort', this.sort_changed, this);
+        this.model.on('change:filters', this.filters_changed, this);
+        this.model.on('change:plugin_config', this.plugin_config_changed, this);
+        this.model.on('change:dark', this.dark_changed, this);
+
+        // Watch the viewer DOM so that widget state is always synchronized with DOM attributes.
+        const observer = new MutationObserver(this._synchronize_state.bind(this));
+        observer.observe(this.pWidget.viewer, {
+             attributes: true,
+             attributeFilter: ["plugin", "columns", "row-pivots", "column-pivots", "aggregates", "sort", "filters", "computed_column"],
```
Comment (timkpaine) [inline packages/perspective-jupyterlab/src/ts/view.ts]: Computed_column should match others even though not implemented in python yet

File: python/perspective/perspective/core/__init__.py [hunk 2]
```diff
@@ -6,9 +6,8 @@
 # the Apache License 2.0.  The full license can be found in the LICENSE file.
 #
 from ._version import __version__  # noqa: F401
-from .base import PerspectiveBaseMixin  # noqa: F401
-from .psp import psp  # noqa: F401
 from .plugin import Plugin  # noqa: F401
 from .aggregate import Aggregate  # noqa: F401
 from .exception import PerspectiveError  # noqa: F401
+from .start import start  # noqa: F401
```
Comment (timkpaine) [inline python/perspective/perspective/core/__init__.py]: Rename

File: python/perspective/perspective/core/start.py [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+#
+
+import pandas
+from functools import wraps
+from ..table.data.pandas import deconstruct_pandas
+from .widget import PerspectiveWidget
+
+
+@wraps(PerspectiveWidget.__init__)
+def start(table_or_data,
```
Comment (timkpaine) [inline python/perspective/perspective/core/start.py]: Rename

File: python/perspective/perspective/core/widget.py [hunk 4]
```diff
@@ -5,103 +5,216 @@
...
 #
+import json
+from datetime import datetime
+from random import random
 from ipywidgets import Widget
-from traitlets import Unicode
-from .base import PerspectiveBaseMixin
-from .data import type_detect
+from traitlets import observe, Unicode
+from .traitlets import PerspectiveTraitlets
```
Comment (timkpaine) [inline python/perspective/perspective/core/widget.py]: Rename traitlets file, can cause issues on earlier python versions

File: python/perspective/perspective/core/widget.py [hunk 5]
```diff
@@ -5,103 +5,216 @@
...
-    ############
-    # Required #
-    ############
+class DateTimeEncoder(json.JSONEncoder):
+    '''Create a custom JSON encoder that allows serialization of datetime and date objects.'''
+
+    def default(self, obj):
+        if isinstance(obj, datetime):
+            # Convert to milliseconds - perspective.js expects millisecond timestamps, but python generates them in seconds.
+            return obj.timestamp() * 1000
```
Comment (timkpaine) [inline python/perspective/perspective/core/widget.py]: Python2 incompatible

File: python/perspective/perspective/core/widget.py [hunk 6]
```diff
@@ -5,103 +5,216 @@
...
     _view_module_version = Unicode('^0.3.0').tag(sync=True)
-    ############
-
-    def delete(self): self.send({'type': 'delete'})
-
-    def update(self, data): self.send({'type': 'update', 'data': type_detect(data).data})
-
-    def __del__(self): self.send({'type': 'delete'})
 
     def __init__(self,
```
Comment (timkpaine) [inline python/perspective/perspective/core/widget.py]: Allow data

File: python/perspective/perspective/core/widget.py [hunk 7]
```diff
@@ -160,6 +166,7 @@ def load(self, table_or_data, **options):
             new_columns = sorted(table.columns())
 
             if str(new_columns) != str(old_columns):
+                print("New dataset has different columns - resetting widget state.")
```
Comment (timkpaine) [inline python/perspective/perspective/core/widget.py]: use `logging` instead of print

File: python/perspective/perspective/core/widget.py [hunk 8]
```diff
@@ -5,103 +5,234 @@
...
+        >>> widget.sort.append(["a", "asc"])
+        >>> widget.sort
+        [["b", "desc"], ["a", "asc"]]
+        >>> tbl.update({"a": [4, 5]}) # updates to the table reflect on the widget
+    '''
+
+    # Required by ipywidgets for proper registration of the backend
     _model_name = Unicode('PerspectiveModel').tag(sync=True)
     _model_module = Unicode('@finos/perspective-jupyterlab').tag(sync=True)
     _model_module_version = Unicode('^0.3.0').tag(sync=True)
```
Comment (timkpaine) [inline python/perspective/perspective/core/widget.py]: update to `^0.3.9`

File: python/perspective/perspective/core/widget.py [hunk 9]
```diff
@@ -5,103 +5,234 @@
...
     _view_name = Unicode('PerspectiveView').tag(sync=True)
     _view_module = Unicode('@finos/perspective-jupyterlab').tag(sync=True)
     _view_module_version = Unicode('^0.3.0').tag(sync=True)
```
Comment (timkpaine) [inline python/perspective/perspective/core/widget.py]: update to `^0.3.9`

**Comment by texodus** [comment]
Thanks for the PR!  Very cool, also thanks @timkpaine for the extensive review.

**Pull request merged**